File: 2271.maximum-white-tiles-covered-by-a-carpet.py
#
# @lc app=leetcode id=2271 lang=python3
#
# [2271] Maximum White Tiles Covered by a Carpet
#

# @lc code=start
class Solution:
    def maximumWhiteTiles(self, tiles: list[list[int]], carpetLen: int) -> int:
        #Sort the tiles according to the beginning index of each tile
        tiles.sort(key=lambda x: x[0])
        right = ans = cover = 0
        n = len(tiles)
        for left in range(n):
            while right < n and tiles[right][1] < tiles[left][0] + carpetLen - 1:
                cover += tiles[right][1] - tiles[right][0] + 1
                right += 1
            if right < n and tiles[right][0] <= tiles[left][0] + carpetLen - 1:
                ans = max(ans, cover + tiles[left][0] + carpetLen - tiles[right][0])
            else:
                ans = max(ans, cover)
            cover -= tiles[left][1] - tiles[left][0] + 1
        return ans
# @lc code=end

File: 2271.maximum-white-tiles-covered-by-a-carpet.java
/*
 * @lc app=leetcode id=2271 lang=java
 *
 * [2271] Maximum White Tiles Covered by a Carpet
 */
import java.util.Arrays;
// @lc code=start
class Solution {
    public int maximumWhiteTiles(int[][] tiles, int carpetLen) {
        Arrays.sort(tiles, (a, b) -> a[0] - b[0]);
        int n = tiles.length, left = 0, right = 0, sum = 0, ans = 0;
        while(left <= right && right < n){
            //initialize the carpet end points
            int leftBound = tiles[left][0], rightBound = leftBound + carpetLen -1;
            
            if(tiles[right][1] <= rightBound){
                sum += tiles[right][1] - tiles[right][0] + 1;
                right++;
                ans = Math.max(sum, ans);
            }
            else{
                if(rightBound >= tiles[right][0]){
                    ans = Math.max(sum + rightBound - tiles[right][0] + 1, ans);
                }
                sum -= tiles[left][1] - tiles[left][0] + 1;
                left++;
            }
        }
        return ans;
    }
}
// @lc code=end

